feat(conversation api): add streaming + tool call support

sicoyle opened the pull request:
# Description

this pr is a duplicate of https://github.com/dapr/dapr/pull/8790 I have to open another pr as i don't have perms in filinto's fork and he is on vacation so i have to work in a diff pr unfortunately

To make things easier picking this task up I had to go through and understand things. Posting in case this is helpful to others:
<img width="1718" height="517" alt="image" src="https://github.com/user-attachments/assets/e369b164-1518-4bdf-a1a2-281fa1a0e9ad" />
Key:
- Blue is what i need to look at next
- pink is tool related
- red is a rename i want to make
- yellow is memo to self
- black/grey is the flow


## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [ ] Code compiles correctly
- [ ] Created/updated tests
- [ ] Unit tests passing
- [ ] End-to-end tests passing
- [ ] Extended the documentation / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_
- [ ] Specification has been updated / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_
- [ ] Provided sample for the feature / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_

sicoyle commented:
fixing build now

JoshVanL commented on pkg/api/universal/conversation.go:
```diff
@@ -115,26 +110,89 @@ func (a *Universal) ConverseAlpha1(ctx context.Context, req *runtimev1pb.Convers
+		return err
+	}
+
+	// Process streaming request
+	ctx := stream.Context()
+	start := time.Now()
+	var usage *runtimev1pb.ConversationUsage
+	var retErr error // used in defer to track error
+	defer func() {
+		elapsed := diag.ElapsedSince(start)
+		var promptTokens, completionTokens int64
+		if usage != nil {
+			promptTokens = int64(usage.GetPromptTokens())         //nolint:gosec // Intentional use of int64 for backward compatibility
+			completionTokens = int64(usage.GetCompletionTokens()) //nolint:gosec // Intentional use of int64 for backward compatibility
+		}
+		diag.DefaultComponentMonitoring.ConversationInvoked(ctx, req.GetName(), retErr == nil, elapsed, diag.StreamingConversation, promptTokens, completionTokens)
+	}()
+
+	// we use non-streaming scrubber on the non-streamingoutput
+	// we get usage to use in monitoring
+	usage, retErr = pipelineImpl.ProcessStream(ctx, stream, component, componentRequest, req.GetName(), outputScrubber)
+	return retErr
```
Please move the sending process into a new func. Assigning vars (`usage`) which are re-referenced in a `defer` mid function is confusing to read.

JoshVanL commented on pkg/conversation/consts.go:
```diff
@@ -0,0 +1,25 @@
+package conversation
+
+const (
+	// SimulatedStreamingChunkSize is the default chunk size (bytes) for simulated streaming conversation data where
+	// the entire response is sent in chunks to simulate a streaming response and allow clients to process data as it arrives
+	// even if the model does not support streaming natively.
+	SimulatedStreamingChunkSize = 1 << 12 // 4 KiB
```
Why are we choosing this arbitrary number, and why are we calling this "simulated"?

JoshVanL commented on pkg/conversation/consts.go:
```diff
@@ -0,0 +1,25 @@
+	// even if the model does not support streaming natively.
+	SimulatedStreamingChunkSize = 1 << 12 // 4 KiB
+
+	// PIIScrubberStreamingWindowSize is the default window size (bytes) for PII scrubbing while streaming conversation data.
+	// this allows for a look-ahead window of stream data to be scrubbed
+	PIIScrubberStreamingWindowSize = 128 // 128 bytes compromise between performance, memory usage and latency.
```
`// 128 bytes compromise between performance, memory usage and latency.`

How was this determined?

sicoyle replied:
no idea tbh

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+package conversation
+
+import (
+	piiscrubber "github.com/aavaz-ai/pii-scrubber"
+	contribConverse "github.com/dapr/components-contrib/conversation"
+	"github.com/dapr/dapr/pkg/messages"
+	runtimev1pb "github.com/dapr/dapr/pkg/proto/runtime/v1"
+	"github.com/dapr/kit/ptr"
```
Organize imports.

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	// Safely convert uint64 to uint32 for token counts
+	// Token counts realistically never exceed uint32 range (~4.3 billion)
+	return &runtimev1pb.ConversationUsage{
+		PromptTokens:     ptr.Of(uint32(resp.Usage.PromptTokens)),
+		CompletionTokens: ptr.Of(uint32(resp.Usage.CompletionTokens)),
+		TotalTokens:      ptr.Of(uint32(resp.Usage.TotalTokens)),
```
Why do we have a discrepancy between typing?

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+// NeedsInputScrubber checks if PII scrubbing is needed for any input
+func NeedsInputScrubber(req *runtimev1pb.ConversationRequest) bool {
+	for _, input := range req.GetInputs() {
+		if input.GetScrubPII() {
+			return true
+		}
+	}
+	return false
+}
```
Surely we should only be doing the scrubbing on inputs where it is enabled? 
This is returning true for all (?).

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	return false
+}
+
+// NeedsOutputScrubber checks if PII scrubbing is needed for outputs
+func NeedsOutputScrubber(req *runtimev1pb.ConversationRequest) bool {
+	return req.GetScrubPII()
+}
```
Redundant
```suggestion
```

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	return protoToolCalls
+}
+
+// extractToolCallIDFromParts extracts tool call ID from tool result parts
+func extractToolCallIDFromParts(parts []*runtimev1pb.ContentPart) string {
+	for _, part := range parts {
+		if toolResult := part.GetToolResult(); toolResult != nil {
+			return toolResult.GetToolCallId()
+		}
+	}
+	return ""
+}
+
+// extractToolNameFromParts extracts tool name from tool result parts
+func extractToolNameFromParts(parts []*runtimev1pb.ContentPart) string {
+	for _, part := range parts {
+		if toolResult := part.GetToolResult(); toolResult != nil {
+			return toolResult.GetName()
+		}
+	}
+	return ""
+}
```
Why do we not error if it is not found? Why is only one "`part`" have a tool name?

sicoyle replied:
i just deleted this

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	return protoToolCalls
+}
+
+// extractToolCallIDFromParts extracts tool call ID from tool result parts
+func extractToolCallIDFromParts(parts []*runtimev1pb.ContentPart) string {
+	for _, part := range parts {
+		if toolResult := part.GetToolResult(); toolResult != nil {
+			return toolResult.GetToolCallId()
+		}
+	}
+	return ""
+}
```
Same here- why no error, and why do we only return _one_ tool id?

sicoyle replied:
no idea

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+	return ""
+}
+
+// convertComponentsContribPartsToProto converts components-contrib content parts to protobuf format
+func convertComponentsContribPartsToProto(result string) []*runtimev1pb.ContentPart {
+	// Add text content if present
+	if result != "" {
+		return []*runtimev1pb.ContentPart{
+			{
+				ContentType: &runtimev1pb.ContentPart_Text{
+					Text: &runtimev1pb.TextContent{
+						Text: result,
+					},
+				},
+			},
+		}
+	}
+
+	return nil
+}
```
What is this doing?

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+			}
+		} else {
+			// Fallback to legacy content processing (BACKWARD COMPATIBILITY)
+			msg = i.GetContent() //nolint:staticcheck // Intentional use of deprecated field for backward compatibility
+
+			if i.GetScrubPII() && scrubber != nil {
+				scrubbed, err := scrubber.ScrubTexts([]string{i.GetContent()}) //nolint:staticcheck // Intentional use of deprecated field for backward compatibility
+				if err != nil {
+					return nil, messages.ErrConversationInvoke.WithFormat(req.GetName(), err.Error())
+				}
+				msg = scrubbed[0]
+			}
+		}
+
+		// Convert protobuf parts directly to components-contrib content parts
+		var parts []contribConverse.ContentPart
+
+		if len(i.GetParts()) > 0 {
+			// Convert protobuf parts to components-contrib parts directly
+			parts = convertProtoPartsToComponentsContrib(i.GetParts())
+		} else if msg != "" {
+			// Legacy mode: convert string content to text part
+			parts = append(parts, contribConverse.TextContentPart{Text: msg})
+		}
+
+		// Create conversation input with content parts
+		inputs = append(
+			inputs,
+			contribConverse.ConversationInput{
+				Message: msg, // Keep for backward compatibility. TODO: remove this field
+				Role:    role,
+				Parts:   parts,
```
Please create new message types.

JoshVanL commented on pkg/conversation/conversation.go:
```diff
@@ -0,0 +1,356 @@
+				toolName := extractToolNameFromParts([]*runtimev1pb.ContentPart{part})
+
+				// Handle tool result messages - determine role based on tool call ID and name
+				if toolCallID != "" && toolName != "" && (role == contribConverse.RoleUser || role == "") {
```
These fields should be marked as optional (?).

JoshVanL commented on pkg/conversation/middlewares.go:
```diff
@@ -0,0 +1,211 @@
+
+**BUG**: When IP addresses and credit card numbers appear in the same text, the library
+fails to detect credit card numbers. This affects ALL usage of the library, not just streaming.
+
+**EVIDENCE**:
+- ✅ "Card number 4263982640269299" → "Card number <CREDIT_CARD>" (WORKS)
+- ❌ "IP is 192.168.1.100 and card 4263982640269299" → "IP is <IP> and card 4263982640269299" (FAILS)
```
Please do not use emoji art in code comments.

JoshVanL commented on pkg/conversation/middlewares.go:
```diff
@@ -0,0 +1,211 @@
+- ✅ "Card number 4263982640269299" → "Card number <CREDIT_CARD>" (WORKS)
+- ❌ "IP is 192.168.1.100 and card 4263982640269299" → "IP is <IP> and card 4263982640269299" (FAILS)
+
+**WORKAROUNDS IMPLEMENTED**:
```
Why is this a "middleware"?

JoshVanL commented on pkg/conversation/middlewares.go:
```diff
@@ -0,0 +1,211 @@
+	return unicode.IsSpace(r) || unicode.IsPunct(r) || unicode.IsSymbol(r)
+}
+
+// findLastCompleteWordPosition finds the position after the last complete word
+// Returns -1 if no complete word boundary found
+func findLastCompleteWordPosition(data []byte) int {
+	if len(data) == 0 {
+		return -1
+	}
+
+	// Find the last separator that creates a complete word boundary
+	for i := len(data) - 1; i >= 0; i-- {
+		if isTokenDelimiter(rune(data[i])) {
+			// Check if there's non-delimiter content after this position
+			hasContentAfter := false
+			for j := i + 1; j < len(data); j++ {
+				if !isTokenDelimiter(rune(data[j])) {
+					hasContentAfter = true
+					break
+				}
+			}
+
+			// If no content after delimiter, it's a good word boundary
+			if !hasContentAfter {
+				return i + 1
+			}
+		}
+	}
+
+	return -1
+}
```
What is the intention of this doing?

JoshVanL commented on pkg/conversation/scrub_streaming_test.go:
```diff
@@ -0,0 +1,1153 @@
+/*
```
Do not use underscores in file names (except with `_test.go`).

sicoyle replied:
and mock based on the rest of the repo :)

JoshVanL commented on pkg/conversation/scrub_streaming_test.go:
```diff
@@ -0,0 +1,1153 @@
+**BUG**: When IP addresses and credit card numbers appear in the same text, the library
+fails to detect credit card numbers. This affects ALL usage of the library.
+
+**EVIDENCE**:
+- ✅ "Card number 4263982640269299" → "Card number <CREDIT_CARD>" (WORKS)
+- ❌ "IP is 192.168.1.100 and card 4263982640269299" → "IP is <IP> and card 4263982640269299" (FAILS)
+
```
Please do not use emoji art in code comments- it's distracting.

JoshVanL commented on pkg/conversation/middlewares.go:
```diff
@@ -0,0 +1,211 @@
+	if len(chunk) == 0 {
+		return chunk, nil
+	}
+
+	s.buffer = append(s.buffer, chunk...)
+
+	// Buffer until window size is reached
+	if len(s.buffer) <= s.windowSize {
+		return nil, nil
+	}
+
+	// Find the last complete word within the window
+	windowContent := s.buffer[:s.windowSize]
+	lastCompletePos := findLastCompleteWordPosition(windowContent)
+
+	if lastCompletePos == -1 {
+		// No complete words - prevent infinite buffering
+		if len(s.buffer) > s.windowSize*2 {
```
Why are we using flushers, manual buffers etc. over actual `io` streams?

sicoyle replied:
after looking into the existing code on this, it appears that the sdk chosen does not support streaming I/O and also honestly we need to get rid of this pkg imo bc it has a bug according to the docs in this file where it fails to detect credit card numbers when IP addresses are present in the same text 💀

JoshVanL commented on pkg/conversation/scrub_streaming_test.go:
```diff
@@ -0,0 +1,1153 @@
+					cardMissed := strings.Contains(result, "4263982640269299")
+					ipMissed := strings.Contains(result, "192.168.1.100")
+
+					t.Logf("📊 DISCOVERY MODE:")
+					if cardMissed {
+						t.Logf("   Credit card: ❌ MISSED (4263982640269299 still visible)")
+					} else {
+						t.Logf("   Credit card: ✅ SCRUBBED")
+					}
+
+					if ipMissed {
+						t.Logf("   IP address:  ❌ MISSED (192.168.1.100 still visible)")
+					} else {
+						t.Logf("   IP address:  ✅ SCRUBBED")
+					}
+
```
Remove all `t.Log` from tests.

JoshVanL commented on pkg/conversation/scrub_streaming_test.go:
```diff
@@ -0,0 +1,1153 @@
+						t.Logf("   IP address:  ✅ SCRUBBED")
+					}
+
+					if cardMissed && !ipMissed {
+						t.Logf("🔍 PATTERN: IP scrubbed, card missed (same as IP-first case)")
+					} else if !cardMissed && ipMissed {
+						t.Logf("🔍 PATTERN: Card scrubbed, IP missed (opposite of IP-first case)")
+					} else if cardMissed && ipMissed {
+						t.Logf("🔍 PATTERN: Both missed (mutual interference)")
+					} else {
+						t.Logf("🔍 PATTERN: Both scrubbed (order matters - this works!)")
+					}
```
Is this unit test actually checking anything?

It seems clear that this file has been written with AI- but it doesn't look like the AI is actually asserting any test cases???

JoshVanL commented on pkg/conversation/scrub_streaming_test.go:
```diff
@@ -0,0 +1,1153 @@
+
+		t.Logf("✅ SUCCESS: Pipeline properly handles error returns from middleware")
+	})
+}
```
It's impossible to review this file.

JoshVanL commented on pkg/conversation/streaming.go:
```diff
@@ -0,0 +1,352 @@
+
+// StreamingMiddleware interface for processing streaming chunks with error handling
+type StreamingMiddleware interface {
+	ProcessChunk(chunk []byte) ([]byte, error)
```
Use proper Go `io` streams when processing streams of data.

JoshVanL commented on pkg/proto/runtime/v1/dapr_tracing.go:
```diff
@@ -88,123 +88,163 @@ func (*DeleteJobRequest) AppendSpanAttributes(rpcMethod string, m map[string]str
 func (*DecryptRequest) AppendSpanAttributes(rpcMethod string, m map[string]string) {
```
Please revert this file.

JoshVanL commented on pkg/diagnostics/component_monitoring.go:
```diff
@@ -31,6 +31,13 @@ const (
+type ConversationType string
+
+const (
+	NonStreamingConversation ConversationType = "non_streaming"
+	StreamingConversation    ConversationType = "streaming"
```
Please use `unary` or some other verbiage to describe this mode, rather than the negation of the other mode.

JoshVanL commented on pkg/scheduler/server/internal/etcd/etcd.go:
```diff
@@ -72,25 +79,53 @@ type etcd struct {
 
 func New(opts Options) (Interface, error) {
-	config, err := config(opts)
+	if opts.Embed {
```
Changes not related to PR.

sicoyle replied:
i rebased onto main and fixed this

JoshVanL commented on tests/integration/framework/process/scheduler/etcd/etcd.go:
```diff
@@ -0,0 +1,131 @@
+/*
```
Stopping review now as there seems to be some broken history or cherry pick causing the diff to become wrong.

JoshVanL reviewed (changes requested):
First pass from me

sicoyle commented on dapr/proto/runtime/v1/dapr.proto:
```diff
@@ -1329,32 +1332,145 @@ message ConversationRequest {
+  string tool_call_id = 1;
+  string name = 2;
+  string content = 3;                 // Tool result as text
+  optional bool is_error = 4;         // Indicates tool execution error
```
good call out. If a tool call is a small part to a larger conversation, and the overall RPC is successful then it is fine to encode tool errors within the message, not as an RPC err. RPC errs should be reserved for transport/protocol level failures across the entire conversation request failing not just a single tool call that is part of the greater conversation. I'll update this to be more like
```
message ConversationToolResult {
  string tool_call_id = 1;
  string name = 2;
  string content = 3;
  oneof result {
    string output_text = 3;
    ToolError error = 4;
  }
}

message ToolError {
  string message = 1;
  int32 code = 2;
}
```

JoshVanL replied:
Please can you tell me why an error from a tool is not an RPC error, and is instead part of the message?

sicoyle commented on dapr/proto/runtime/v1/dapr.proto:
```diff
@@ -1329,32 +1332,145 @@ message ConversationRequest {
+  string id = 1;
+  // The type of tool call (e.g., "function")
+  string type = 2;
+  // Name of the function to call
+  string name = 3;
```
Tools don't have to just be a function, so leaving this. But I can think on how to make this cleaner 🤷‍♀️

sicoyle commented:
I pushed where I'm at so far with addressing feedback and doing cleanup, renaming, etc. Check out my PR description image for a better understanding if you need - it helped me a ton doing that with following the names and whatnot. Also pushed on contrib corresponding changes. That PR needs quite a bit of love too tbh. I put comments in code here with `TODO(<at my github tag>)` that could use some input @JoshVanL so far, and I will work on addressing the rest of your existing comments tomorrow.

sicoyle commented on dapr/proto/runtime/v1/dapr.proto:
```diff
@@ -1329,32 +1332,145 @@ message ConversationRequest {
+  optional uint32 completion_tokens = 2 [json_name = "completionTokens"];
+  // Total number of tokens used
+  optional uint32 total_tokens = 3 [json_name = "totalTokens"];
+}
```
This was an add in to bubble up token usage with the llm provider api 🤷‍♀️

sicoyle commented:
closing to do tool calling only for now
